addon/PreviewProvider.js

```diff
@@ -208,38 +207,93 @@ PreviewProvider.prototype = {
+      const {favicon_url: tippyTopFavicon, background_color: tippyTopBackgroundColor, metadata_source} = this._tippyTopProvider.processSite(link);
+      enhancedLink.metadata_source = metadata_source;
 
       // Find the item in the map and return it if it exists, then unpack that
       // object onto our new link
-      if (existingLinks.has(site.cache_key)) {
-        const cachedMetadataLink = existingLinks.get(site.cache_key);
+      let metadataLinkFavicons = null;
+      if (existingLinks.has(link.cache_key)) {
+        const cachedMetadataLink = existingLinks.get(link.cache_key);
+        metadataLinkFavicons = cachedMetadataLink.favicons;
+        enhancedLink.metadata_source = cachedMetadataLink.metadata_source;
         enhancedLink.title = cachedMetadataLink.title;
         enhancedLink.description = cachedMetadataLink.description;
         enhancedLink.provider_name = cachedMetadataLink.provider_name;
         enhancedLink.images = cachedMetadataLink.images;
-        enhancedLink.favicons = cachedMetadataLink.favicons;
-        enhancedLink.metadata_source = cachedMetadataLink.metadata_source;
-        if (cachedMetadataLink.favicons && cachedMetadataLink.favicons.length) {
-          enhancedLink.favicon_url = cachedMetadataLink.favicons[0].url;
-        }
       }
 
-      // Add tippy top data, if available
-      Object.assign(enhancedLink, this._tippyTopProvider.processSite(enhancedLink));
-
-      return enhancedLink;
-    }).filter(link => link);
+      // consolidate favicons, background color and metadata source, then return the final link
+      enhancedLink.favicon_url = this._consolidateFavicons(tippyTopFavicon, metadataLinkFavicons, firefoxFavicon);
+      enhancedLink.background_color = this._consolidateBackgroundColors(tippyTopBackgroundColor, metadataLinkFavicons, firefoxBackgroundColor);
+      results.push(enhancedLink);
+    }
 
     this._tabTracker.handlePerformanceEvent(event, "previewCacheHits", existingLinks.size);
     this._tabTracker.handlePerformanceEvent(event, "previewCacheMisses", processedLinks.length - existingLinks.size);
     return results;
   }),
 
+  /*
+   * Consolidate favicons from tippytop, firefox, and our own metadata. Tippytop
+   * is our number 1 choice, followed by the favicon returned by the metadata service,
+   * and finally firefox's data URI as a last resort
+   */
+  _consolidateFavicons(tippyTopFavicon, metadataFavicons, firefoxFavicon) {
+    const metadataFavicon = metadataFavicons && metadataFavicons[0] && metadataFavicons[0].url;
+    return tippyTopFavicon || metadataFavicon || firefoxFavicon;
+  },
+
+  /*
+   * Consolidate the background colors from tippytop, firefox, and our own metadata.
+   * TippyTop background color and metadata background color both need to be converted
+   * from hex to an rgb array, whereas the firefox background color is already rgb
+   */
+  _consolidateBackgroundColors(tippyTopBackgroundColor, metadataFavicons, firefoxBackgroundColor) {
```

Comment:
personally I would probably put this and `hexToRgb` in a separate file in `addons/lib` perhaps – it would make them easier to test/isolate all the different cases (with the background fade, etc.) If it's a lot of extra work don't worry about it too much though.